### PR TITLE
Explicitly declare our build backend

### LINKED ISSUE #1644: Tests failing to run with BackendUnavailable since pip 19
It seems that since the release of pip 19, Setuptools' own tests fail to run thus:

```
setuptools fix_889_and_non-ascii_in_setup.cfg_take_2 $ tox -e py27                                                                                                                       
py27 create: /Users/jaraco/code/main/setuptools/.tox/py27
py27 installdeps: -rtests/requirements.txt
py27 develop-inst: /Users/jaraco/code/main/setuptools
ERROR: invocation failed (exit code 2), logfile: /Users/jaraco/code/main/setuptools/.tox/py27/log/py27-2.log
ERROR: actionid: py27
msg: developpkg
cmdargs: '/Users/jaraco/code/main/setuptools/.tox/py27/bin/pip install --exists-action w -e /Users/jaraco/code/main/setuptools'

DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.
Obtaining file:///Users/jaraco/code/main/setuptools
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
Exception:
Traceback (most recent call last):
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/cli/base_command.py", line 176, in main
    status = self.run(options, args)
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/commands/install.py", line 315, in run
    resolver.resolve(requirement_set)
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/resolve.py", line 131, in resolve
    self._resolve_one(requirement_set, req)
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/resolve.py", line 294, in _resolve_one
    abstract_dist = self._get_abstract_dist_for(req_to_install)
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/resolve.py", line 226, in _get_abstract_dist_for
    req, self.require_hashes, self.use_user_site, self.finder,
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/operations/prepare.py", line 382, in prepare_editable_requirement
    abstract_dist.prep_for_dist(finder, self.build_isolation)
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_internal/operations/prepare.py", line 149, in prep_for_dist
    reqs = self.req.pep517_backend.get_requires_for_build_wheel()
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_vendor/pep517/wrappers.py", line 71, in get_requires_for_build_wheel
    'config_settings': config_settings
  File "/Users/jaraco/code/main/setuptools/.tox/py27/lib/python2.7/site-packages/pip/_vendor/pep517/wrappers.py", line 162, in _call_hook
    raise BackendUnavailable
BackendUnavailable

py27 installed: DEPRECATION: Python 2.7 will reach the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 won't be maintained after that date. A future version of pip will drop support for Python 2.7.,apipkg==1.5,atomicwrites==1.2.1,attrs==18.2.0,configparser==3.7.1,contextlib2==0.5.5,coverage==4.5.2,enum34==1.1.6,execnet==1.5.0,flake8==3.6.0,funcsigs==1.0.2,futures==3.2.0,importlib-metadata==0.8,mccabe==0.6.1,mock==2.0.0,more-itertools==5.0.0,path.py==11.5.0,pathlib2==2.3.3,pbr==5.1.1,pluggy==0.8.1,py==1.7.0,pycodestyle==2.4.0,pyflakes==2.0.0,pytest==3.10.1,pytest-cov==2.6.1,pytest-fixture-config==1.4.0,pytest-flake8==1.0.3,pytest-shutil==1.4.0,pytest-virtualenv==1.4.0,scandir==1.9.0,six==1.12.0,termcolor==1.1.0,virtualenv==16.3.0,zipp==0.3.3
________________________________________________________________________________________ summary ________________________________________________________________________________________
ERROR:   py27: InvocationError for command /Users/jaraco/code/main/setuptools/.tox/py27/bin/pip install --exists-action w -e /Users/jaraco/code/main/setuptools (see /Users/jaraco/code/main/setuptools/.tox/py27/log/py27-2.log) (exited with code 2)
```

I suspect this is related to #1642 if not just another manifestation of the same issue.

### PR DESCRIPTION
**Pull request by pganssle**
## Summary of changes
This change more accurately reflects that ``setuptools`` is intended to use the ``build_meta`` backend from a previous installation. This may cause some problems with source builds if existing wheels for ``wheel`` and ``setuptools`` are not available. In that event, manually building a source wheel is recommended.

This may break `pip install --no-binary :all:`, though it's not entirely clear that it will. I would take the position that that this is basically revealing a bug in `pip`, but we could also just call this a "breaking change".

Closes #1644

### Pull Request Checklist
- [ ] News fragment added in changelog.d. See [documentation](http://setuptools.readthedocs.io/en/latest/developer-guide.html#making-a-pull-request) for details

### CODE REVIEW

**Comment by pganssle** [comment]
@pfmoore @pradyunsg Does `pip` actually need to do anything for this case, or have we worked out that this won't be a problem?

**Comment by pganssle** [comment]
@pradyunsg So you are saying you're fine with `--no-binary :all:` breaking?

**Comment by pfmoore** [comment]
I think we need to resolve the "backend bootstrapping" discussion on Discourse before this can go in. Once we have a solution there (which may involve changes to pip and/or modifications to what you're doing here) then this can go in.

For now, my preference would be to put this on hold.

**Comment by pganssle** [comment]
@pfmoore I think you can see my reasoning on #1644, but I am fairly certain that that discourse thread is going nowhere, certainly nowhere useful. From where I'm standing, the spec should allow for this. If a frontend (`pip`) cannot in some situations provide the `setuptools` build dependencies, that seems like a problem with that frontend, not with our configuration.

**Comment by pganssle** [comment]
The test failures seem to actually be due to pypa/pip#6197, so I think we can pin `pip!=19.0.0,!=19.0.1` in the tests.

**Comment by pfmoore** [comment]
I'm not willing to get stuck in a confrontation over this, and in particular I'm not going to take a view on "whose problem is it?" while we're trying to establish agreement on what the responsibilities are over on Discourse (I don't share your disillusionment with that thread, although it'll be hard to make progress if setuptools isn't engaged with the process :-()

All I'm going to say here is that a cycle in build requirements is **logically impossible** to resolve if there's no binaries available. So whatever pip does, this change forces a bootstrapping step on anyone starting from a "sources only" environment. I'm not going to judge whether such environments constitute valid use cases, but conversely, if people raise bug reports on pip in that situation, I'll probably  point the reporters at setuptools for advice. If people are using `--no-binary :all:`, that's a bit different, as wheels potentially *are* available and they have chosen to disallow them. In that case I'd say that setuptools build configuration doesn't work with `--no-binary :all:` and we're working on a resolution, but we don't yet know whether that resolution will need changes in pip or in setuptools.

That's a complex message to get across, and I'd prefer to avoid the need by postponing this change. But it's not my place to dictate what setuptools does, so if you want to go ahead we can work with the consequences.

**Comment by pganssle** [comment]
> All I'm going to say here is that a cycle in build requirements is **logically impossible** to resolve if there's no binaries available. So whatever pip does, this change forces a bootstrapping step on anyone starting from a "sources only" environment.

This is certainly true, though it's also true for `gcc` and `clang` and they seem to get along just fine. The kind of people who enforce "sources only" environments are also, generally, the kind of people who know how to deal with it. I think we can put it in a FAQ somewhere, maybe in packaging.python.org is the most appropriate place.

We can probably put this off for some time, but at *some point* we'll have to call the discourse thread a waste of time and give up on it. Maybe that's after the discussion peters out or maybe it's because enough people get tired of the discussion that the only people left are the ones who can agree on a path forward.

That said, if PEP 517 changes to force bootstrap builds, I imagine the "always build from source" crowd might like that there's a version of `setuptools` that supports PEP 517 and declares a build requirement rather than vendoring a wheel in the `sdist`. It's a lot easier to ban wheels at the infrastructure level than it is to patch the `sdist` with wheels you bootstrap by some other, totally unsupported mechanism. If we postpone the change, there will only be versions of `setuptools` that don't support PEP 517 and versions of `setuptools` that vendor wheels.

**Comment by dstufft** [comment]
I would point out that my proposal that I posted yesterday explicitly allowed setuptools to basically land this exact pull request. It highly suggests *not* landing a PR like this, by saying that backends **SHOULD** avoid introducing a cyclic build dependency graph but it doesn't explicitly disallow it, since it's not a **MUST**.

I would also note that one of the questioned I raised in my proposal was whether we should make it a **SHOULD** or whether we should make it a **MUST**. Obviously if we made it a **MUST** then a PR like this would make setuptools non-compliant with PEP 517 (but obviously the PEP has no teeth to force compliance), in either case, pip will have to detect and reject cyclic build dependencies when attempting to build.

In all cases it would mean that ``pip install --no-binary :all: ...`` would break, and if those people landed on pip's tracker we'd redirect them to setuptools tracker. Although if someone was generally OK with installing everything but setuptools from source, they could . do ``pip install --no-binary :all: --only-binary setuptools ...` instead.

**Comment by pfmoore** [comment]
> `pip install --no-binary :all: --only-binary setuptools ...`

Hmm, I didn't know that `--no-binary` and `--only-binary` interacted that way (i.e., correctly ;-)) That's good to know and means that there's a much easier workaround for people who want to do this than I was assuming.

**Comment by pganssle** [comment]
> In all cases it would mean that `pip install --no-binary :all: ...` would break, and if those people landed on pip's tracker we'd redirect them to setuptools tracker.

I'm not entirely sure why this would be necessary given that this was a deliberate choice and not a bug. I imagine that would just be an exercise in frustration for the user. Presumably if this becomes a problem we can write some documentation to indicate what to do and send users to that, though obviously my preferred solution is something that just works for the users in the first place.

**Comment by dstufft** [comment]
I mean. That’s not like being passive aggressive about it or anything, it’s just saying that if someone comes to pip because the specific choices made by a project are causing them pain, that’s feedback we’d direct towards that project. The same as If a project opted not to upload a sdist at all and the user came to pip saying —no-binary didn’t work for them.

**Comment by pfmoore** [comment]
> my preferred solution is something that just works for the users in the first place.

But if the users say "I want to build everything from source (`--no-binary :all:`)" and setuptools says "I need setuptools and wheel installed in order to install myself", then there *is* no "just works". All we can do is say "you need to have setuptools available as a wheel because it says in its build instructions that it can't build from source without one" and the way to do that is `--no-binary :all: --only-binary setuptools` and provide a wheel (or let pip get it from PyPI). There's no magic here, you get what you say you need, and the user has to provide it.

**Comment by pganssle** [comment]
I ducked out of the discourse thread for a reason, so I'm not really interested in moving that whole discussion onto this random PR. I'm sorry to bring up that I would prefer if we chose a different resolution to the bootstrapping problem.

When we have a satisfactory resolution to #1659, I think we can fix the tests for this PR, then we can decide whether or not to land it. If we land it, I suspect that the most useful thing to do would be to put a note in one or more of: packaging.python.org, setuptools documentation and pip. When people inevitably complain, the best thing to do would be to point to that documentation.

**Comment by pfmoore** [comment]
@pganssle Is there a misunderstanding here? `requires = ["setuptools", "wheel"]` is saying that "in order to build setuptools from source, a frontend needs to create a build environment with setuptools and wheel installed in it. If a user says "don't download wheels from PyPI" (however they spell that), then the only way the frontend can get setuptools installed in that build environment is by building it from source. And that's an infinite regression right there.

I think you're hoping that frontends will be required to use wheels to satisfy build requirements. But we *can't* - it's not even about `--no-binary`, if the user is running the pip in an environment with no access to PyPI and only has sdists locally, they can't get a wheel. So we have to explain to them what's going on, and "setuptools needs setuptools to build itself, so you'll have to seed that process outside of pip" is all we can say.

**Comment by pfmoore** [comment]
@pradyunsg To be fair, I'm OK with `--no-binary :all:` breaking (and I suspect @dstufft is too) but we're saying that when it does we'd probably be saying "... and that's caused by a bug in setuptools", so the fact that we're OK with it breaking probably isn't actually as helpful as @pganssle hopes...

But if he wants to release this change and deal with the results, then that's fine. And I'm also not being passive aggressive by saying this, it's just the reality of how things are right now, that `requires` line will have that result. Once the Discourse thread produces a result, maybe there'll be another alternative.

**Comment by pganssle** [comment]
As to the question of deferring it, I suppose having a release that explicitly declares its backend dependencies vs. including them as wheels in the `sdist` is not terribly important because without a corresponding version of `pip` that handles it smoothly it buys you nothing: when there's a version that has wheels in the `sdist`, it will just "work smoothly" and you won't know you're violating your "everything from source" policy, and if you do use the cyclic dependency version it will break with `pip install --no-binary :all:`. In any case, once you know you have to manually build a setuptools bootstrap wheel, you can patch `setuptools`.

So I suppose there's no advantage in actually rushing this through if `pip` is not going to have a corresponding fix in place. We can wait to see if the problem solves itself before landing this.

**Comment by pganssle** [comment]
> when it does we'd probably be saying "... and that's caused by a bug in setuptools"

This is the problem with the framing: I think it's uncharitable to consider this a "bug" in setuptools. It would be a *deliberate choice* to make explicit that `setuptools` needs a way to bootstrap itself. We could just as easily say it's a "bug" in `pip` that there's no way to express "build everything from sdists if possible and fall back to wheels to break cyclic build-time dependencies".

**Comment by dstufft** [comment]
Assuming that both my proposal is accepted and we use the SHOULD terminology then I wouldn’t consider it a bug in setuptools. I’d just consider it an outcome of the choices made and if people want different choices then they would need to provide that feedback to the setuptools project. 

If we use the MUST terminology then it would be a setuptools bug. 

If we accept a different proposal then who knows, it would depend on the specifics of that proposal.

**Comment by pfmoore** [comment]
Yeah, I wasn't clear. When we have a revision to the PEP, then how we frame the response would depend on the wording on that (and I wouldn't expect it to be a "bug" in anything).

Right now, though, I think it's fair to say that this PR does not do what you want it to in conjunction with the existing behaviour in pip. That's why I advise not releasing it yet, and why I'd say to anyone affected that "a change in setuptools triggered this issue" (which is a better way of wording it than "there's a bug in setuptools").

**Comment by pganssle** [comment]
> Right now, though, I think it's fair to say that this PR does not do what you want it to in conjunction with the existing behaviour in pip. That's why I advise not releasing it yet, and why I'd say to anyone affected that "a change in setuptools triggered this issue" (which is a better way of wording it than "there's a bug in setuptools").

To be clear, this *does* improve the situation, because right now we're not shipping our `pyproject.toml` in the `sdist` at all, despite the fact that it should be part of the source. The reason we can't ship `pyproject.toml` is because we need to opt out of PEP 517, since we provide our own backend. If we started shipping our `pyproject.toml` *without* this change, then the following would break:

```python
pip download -d dist --no-binary :all: setuptools
pip install dist/setuptools*.zip
```

So, basically, if you've gathered a source distribution (which you can do by using `--no-binary :all:` as part of the download requirements), you can install it or build a wheel from it safely with `pip wheel`, satisfying the build-time dependencies with wheels. Without this change, you'd get an error about a missing backend.

**Comment by pganssle** [comment]
Closing this as not consistent with the newest version of PEP 517.
